lib/OpenQA/WebAPI/Controller/Test.pm

```diff
@@ -727,12 +727,21 @@ sub _prepare_job_results ($self, $all_jobs, $limit) {
 }
 
 sub _prepare_groupids ($self) {
-    return [0] unless my $groups = $self->groups_for_globs;
-    return [map { $_->id } @$groups] if @$groups;
-
     my $v = $self->validation;
     $v->optional('groupid')->num(0, undef);
-    return $v->is_valid('groupid') ? $self->every_param('groupid') : undef;
+    $v->optional('not_groupid')->num(0, undef);
+
+    my $groupids = $v->is_valid('groupid') ? $self->every_param('groupid') : [];
+    my $not_groupids = $v->is_valid('not_groupid') ? $self->every_param('not_groupid') : [];
+
+    if (@$not_groupids) {
+        my %not_groupid_hash = map { $_ => 1 } @$not_groupids;
+        $groupids = [grep { !$not_groupid_hash{$_} } @$groupids];
+    }
+
+    return $groupids if @$groupids;
+    return [0] unless my $groups = $self->groups_for_globs;
+    return [map { $_->id } @$groups];
```

Comment:
I'm wondering what the point of this change is. It is just filtering one set of query parameters by another set of query parameters.